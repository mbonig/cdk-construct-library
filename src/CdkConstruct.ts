import { kebab, pascal } from 'case';
import { Project, SampleFile } from 'projen';
import { AwsCdkConstructLibrary } from 'projen/lib/awscdk';
import { NpmAccess } from 'projen/lib/javascript';

export interface CdkConstructOptions {
  /**
   * The name of the construct. Should be formatted @matthewbonig/<name>.
   *
   * @default 'my-construct'
   * @example 'my-construct'
   */
  readonly name: string;

  /**
   * The version of the CDK to use.
   *
   * @default '2.85.0'
   */
  readonly cdkVersion?: string;

  /**
   * The name of the repository. If not specified, the name will be used.
   *
   * @example 'my-test-construct'
   * @default "https://github.com/mbonig/<name>.git"
   */
  readonly repositoryName?: string;

  /**
   * The version of the constructs library to use.
   *
   * @default '10.1.203'
   */
  readonly constructsVersion?: string;

  /**
   * The description of the construct.
   */
  readonly description?: string;

  /**
   * Any additional entries to add to the .gitignore file.
   */
  readonly gitIgnore?: string[];

  /**
   * Additional keywords to include in the package.json file.
   */
  readonly keywords?: string[];

  /**
   * Disable publishing to PyPi.
   *
   * @default false
   */
  readonly disablePublishToPypi?: boolean;

  /**
   * Disable publishing to NuGet.
   *
   * @default false
   */
  readonly disablePublishToNuGet?: boolean;

  /**
   * Disable publishing to Maven.
   *
   * @default false
   */
  readonly disablePublishToMaven?: boolean;

  /**
   * Disable publishing to Go.
   *
   * @default false
   */
  readonly disablePublishToGo?: boolean;
}

class SampleConstructFile extends SampleFile {
  constructor(scope: Project, constructName: string) {
    const propsInterfaceName = `${pascal(constructName)}Props`;
    super(scope, `src/${pascal(constructName)}.ts`, {
      contents: [
        "import { Construct } from 'constructs';",
        '',
        `export interface ${propsInterfaceName} {}`,
        '',
        `export class ${pascal(constructName)} extends Construct {`,
        `  constructor(scope: Construct, id: string, props: ${propsInterfaceName}) {`,
        '    super(scope, id);',
        '  }',
        '}',
      ].join('\n'),
    });
  }
}

class ConstructIndexFile extends SampleFile {
  constructor(scope: Project, constructName: string) {
    super(scope, 'src/index.ts', {
      contents: `export * from \'./${pascal(constructName)}\';`,
    });
  }

}

class TestFile extends SampleFile {
  constructor(scope: Project, constructName: string) {
    super(scope, `test/${pascal(constructName)}.test.ts`, {
      contents: [
        "import { App, Stack } from 'aws-cdk-lib';",
        "import { Template } from 'aws-cdk-lib/assertions';",
        `import { ${pascal(constructName)} } from '../src';`,

        'test(\'Snapshot\', () => {',
        '  // GIVEN',
        '  const stack = new Stack();',
        '',
        '  // WHEN',
        `  new ${pascal(constructName)}(stack, '${pascal(constructName)}', {});`,
        '',
        '  // THEN',
        '  const assert = Template.fromStack(stack);',
        '  expect(assert.toJSON()).toMatchSnapshot();',
        '});',
      ].join('\n'),
    });
  }

}

export class CdkConstruct extends AwsCdkConstructLibrary {

  constructor(options: CdkConstructOptions) {
    super({
      author: 'Matthew Bonig',
      authorAddress: 'matthew.bonig@gmail.com',
      description: options.description,
      cdkVersion: options.cdkVersion || '2.85.0',
      constructsVersion: options.constructsVersion || '10.1.203',
      defaultReleaseBranch: 'main',
      name: `@matthewbonig/${options.name}`,
      repositoryUrl: `https://github.com/mbonig/${options.repositoryName || options.name}.git`,
      license: 'MIT',
      npmAccess: NpmAccess.PUBLIC,
      gitignore: ['.idea/', ...(options?.gitIgnore ?? [])],
      keywords: ['awscdk', 'cdk', ...(options.keywords ?? [])],
      depsUpgrade: false,
      publishToPypi: options.disablePublishToPypi ? undefined : {
        distName: `mbonig.${kebab(options.name)}`,
        module: `mbonig.${kebab(options.name)}`,
      },
      publishToNuget: options.disablePublishToNuGet ? undefined : {
        dotNetNamespace: `MBonig.${pascal(options.name)}`,
        packageId: `MBonig.${pascal(options.name)}`,
      },
      publishToMaven: {
        mavenGroupId: 'com.matthewbonig',
        javaPackage: `com.matthewbonig.${pascal(options.name)}`,
        mavenArtifactId: kebab(options.name),
      },
      publishToGo: options.disablePublishToGo ? undefined : {
        moduleName: 'github.com/mbonig/cdk-go',
        gitUserName: 'Matthew Bonig',
        gitUserEmail: 'matthew.bonig@gmail.com',
      },
      sampleCode: false,
    });

    new SampleConstructFile(this, options.name);
    new ConstructIndexFile(this, options.name);
    new TestFile(this, options.name);
  }
}
