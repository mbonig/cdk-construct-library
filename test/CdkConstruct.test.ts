import { synthSnapshot } from 'projen/lib/util/synth';
import { parse } from 'yaml';
import { CdkConstruct, CdkConstructOptions } from '../src';

describe('CdkConstruct', () => {
  function getSnapshot(options?: Partial<CdkConstructOptions>) {
    const project = new CdkConstruct({
      name: 'test-construct',
      ...options,
    });
    return synthSnapshot(project);
  }

  test('Author is me', () => {
    const snapshot = getSnapshot();
    expect(snapshot['package.json'].author.name).toEqual('Matthew Bonig');
  });

  test('Author address is my email', () => {
    const snapshot = getSnapshot();
    expect(snapshot['package.json'].author.email).toEqual('matthew.bonig@gmail.com');
  });

  test('CDK version defaults to 2.85.0', () => {
    const snapshot = getSnapshot();
    expect(snapshot['package.json'].peerDependencies['aws-cdk-lib']).toEqual('^2.85.0');
  });

  test('CDK version can be overridden', () => {
    const snapshot = getSnapshot({
      cdkVersion: '2.114.0',
    });
    expect(snapshot['package.json'].peerDependencies['aws-cdk-lib']).toEqual('^2.114.0');
  });

  test('package name is correct', () => {
    const snapshot = getSnapshot();
    expect(snapshot['package.json'].name).toEqual('@matthewbonig/test-construct');
  });

  test('repo is correct when not overridden', () => {
    const snapshot = getSnapshot();
    expect(snapshot['package.json'].repository.type).toEqual('git');
    expect(snapshot['package.json'].repository.url).toEqual('https://github.com/mbonig/test-construct.git');
  });

  test('repo is correct when overridden', () => {
    const snapshot = getSnapshot({
      repositoryName: 'my-test-construct',
    });
    expect(snapshot['package.json'].repository.type).toEqual('git');
    expect(snapshot['package.json'].repository.url).toEqual('https://github.com/mbonig/my-test-construct.git');
  });

  test('constructs version is correct when not overridden', () => {
    const snapshot = getSnapshot({});
    expect(snapshot['package.json'].devDependencies.constructs).toEqual('10.1.203');
  });

  test('constructs version is correct when overridden', () => {
    const snapshot = getSnapshot({
      constructsVersion: '10.1.204',
    });
    expect(snapshot['package.json'].devDependencies.constructs).toEqual('10.1.204');
  });

  test('npmAccess is public', () => {
    const snapshot = getSnapshot();
    expect(snapshot['package.json'].publishConfig.access).toEqual('public');
  });

  test('license is MIT', () => {
    const snapshot = getSnapshot();
    expect(snapshot['package.json'].license).toEqual('MIT');
  });

  test('description is used', () => {
    const snapshot = getSnapshot({
      description: 'This is a test',
    });
    expect(snapshot['package.json'].description).toEqual('This is a test');
  });

  test('gitignore includes .idea/', () => {
    const snapshot = getSnapshot();
    expect(snapshot['.gitignore']).toContain('.idea/');
  });

  test('Keywords are added', () => {
    const snapshot = getSnapshot({
      keywords: ['somethingelse'],
    });
    expect(snapshot['package.json'].keywords).toContain('awscdk');
    expect(snapshot['package.json'].keywords).toContain('cdk');
    expect(snapshot['package.json'].keywords).toContain('somethingelse');
  });

  test('depsUpgrade is false by default', () => {
    const snapshot = getSnapshot();
    expect(snapshot['.github/workflows/upgrade-main.yml']).toBeUndefined();
  });

  test('Publishing to PyPi is enabled by default', () => {
    const snapshot = getSnapshot();
    let releaseFile = snapshot['.github/workflows/release.yml'];
    releaseFile = parse(releaseFile);
    expect(releaseFile.jobs.release_pypi).toBeDefined();
  });

  test('Publishing to PyPi has good name', () => {
    const snapshot = getSnapshot();
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.python.distName).toEqual('mbonig.test-construct');
    expect(packageFile.jsii.targets.python.module).toEqual('mbonig.test-construct');
  });

  test('Publishing to PyPi kebab-cases the name', () => {
    const snapshot = getSnapshot({
      name: 'TestConstruct',
    });
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.python.distName).toEqual('mbonig.test-construct');
    expect(packageFile.jsii.targets.python.module).toEqual('mbonig.test-construct');
  });

  test('Can disable publishing to PyPi', () => {
    const snapshot = getSnapshot({
      disablePublishToPypi: true,
    });
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.python).toBeUndefined();
  });

  test('Published to NuGet by default', () => {
    const snapshot = getSnapshot({});
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.dotnet.namespace).toEqual('MBonig.TestConstruct');
    expect(packageFile.jsii.targets.dotnet.packageId).toEqual('MBonig.TestConstruct');
  });

  test('Can disable Published to NuGet', () => {
    const snapshot = getSnapshot({
      disablePublishToNuGet: true,
    });
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.dotnet).toBeUndefined();
  });

  test('Published to Maven by default', () => {
    const snapshot = getSnapshot({});
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.java.package).toEqual('com.matthewbonig.TestConstruct');
    expect(packageFile.jsii.targets.java.maven.groupId).toEqual('com.matthewbonig');
    expect(packageFile.jsii.targets.java.maven.artifactId).toEqual('test-construct');
  });

  test('Can disable Published to Maven', () => {
    const snapshot = getSnapshot({
      disablePublishToMaven: true,
    });
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.maven).toBeUndefined();
  });

  test('Published to Go by default', () => {
    const snapshot = getSnapshot({});
    const packageFile = snapshot['package.json'];


    expect(packageFile.jsii.targets.go.moduleName).toEqual('github.com/mbonig/cdk-go');
  });

  test('Can disable Published to go', () => {
    const snapshot = getSnapshot({
      disablePublishToGo: true,
    });
    const packageFile = snapshot['package.json'];

    expect(packageFile.jsii.targets.go).toBeUndefined();
  });

  test('Creates the starting construct file', () => {
    const snapshot = getSnapshot({});
    const constructFile = snapshot['src/TestConstruct.ts'];

    expect(constructFile).toMatchSnapshot();
  });

  test('Creates the index file', () => {
    const snapshot = getSnapshot({});
    const indexFile = snapshot['src/index.ts'];

    expect(indexFile).toMatchSnapshot();
  });

  test('Creates test file', () => {
    const snapshot = getSnapshot({});
    const testFile = snapshot['test/TestConstruct.test.ts'];

    expect(testFile).toMatchSnapshot();
  });

});
