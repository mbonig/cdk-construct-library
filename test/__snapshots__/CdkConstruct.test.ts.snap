// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdkConstruct Creates test file 1`] = `
"import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { TestConstruct } from '../src';
test('Snapshot', () => {
  // GIVEN
  const stack = new Stack();

  // WHEN
  new TestConstruct(stack, 'TestConstruct', {});

  // THEN
  const assert = Template.fromStack(stack);
  expect(assert.toJSON()).toMatchSnapshot();
});"
`;

exports[`CdkConstruct Creates the index file 1`] = `"export * from './TestConstruct';"`;

exports[`CdkConstruct Creates the starting construct file 1`] = `
"import { Construct } from 'constructs';

export interface TestConstructProps {}

export class TestConstruct extends Construct {
  constructor(scope: Construct, id: string, props: TestConstructProps) {
    super(scope, id);
    const {} = props;
  }
}"
`;
